File: app/models/entity/post.go
package entity

import (
	"fmt"
	"time"

	"github.com/Spicy-Bush/fider-tarkov-community/app/models/enum"
)

// Post represents an post on a tenant board
type Post struct {
	ID                int                   `json:"id"`
	Number            int                   `json:"number"`
	Title             string                `json:"title"`
	Slug              string                `json:"slug"`
	Description       string                `json:"description"`
	CreatedAt         time.Time             `json:"createdAt"`
	LastActivityAt    time.Time             `json:"lastActivityAt"`
	User              *User                 `json:"user"`
	VoteType          int                   `json:"voteType"`
	VotesCount        int                   `json:"votesCount"`
	CommentsCount     int                   `json:"commentsCount"`
	Status            enum.PostStatus       `json:"status"`
	Response          *PostResponse         `json:"response,omitempty"`
	Tags              []string              `json:"tags"`
	TagDates          string                `json:"tagDates,omitempty"`
	LockedSettings    *PostLockedSettings   `json:"lockedSettings,omitempty"`
	ArchivedSettings  *PostArchivedSettings `json:"archivedSettings,omitempty"`
	Upvotes           int                   `json:"upvotes"`
	Downvotes         int                   `json:"downvotes"`
	ModerationPending bool                  `json:"moderationPending,omitempty"`
	ModerationData    string                `json:"moderationData,omitempty"`
}

type PostLockedSettings struct {
	Locked      bool      `json:"locked"`
	LockedAt    time.Time `json:"lockedAt"`
	LockedBy    *User     `json:"lockedBy"`
	LockMessage string    `json:"lockMessage,omitempty"`
}

type PostArchivedSettings struct {
	ArchivedAt     time.Time       `json:"archivedAt"`
	ArchivedBy     *User           `json:"archivedBy"`
	PreviousStatus enum.PostStatus `json:"previousStatus"`
}

// IsLocked returns true if this post is locked
func (p *Post) IsLocked() bool {
	return p.LockedSettings != nil && p.LockedSettings.Locked
}

func (p *Post) IsArchived() bool {
	return p.Status == enum.PostArchived
}

// CanBeVoted returns true if this post can have its vote changed
func (i *Post) CanBeVoted() bool {
	return i.Status != enum.PostCompleted && i.Status != enum.PostDeclined && i.Status != enum.PostDuplicate
}

func (i *Post) Url(baseURL string) string {
	return fmt.Sprintf("%s/posts/%d/%s", baseURL, i.Number, i.Slug)
}

// PostResponse is a staff response to a given post
type PostResponse struct {
	Text        string        `json:"text"`
	RespondedAt time.Time     `json:"respondedAt"`
	User        *User         `json:"user"`
	Original    *OriginalPost `json:"original"`
}

// OriginalPost holds details of the original post of a duplicate
type OriginalPost struct {
	Number int             `json:"number"`
	Title  string          `json:"title"`
	Slug   string          `json:"slug"`
	Status enum.PostStatus `json:"status"`
}

func (i *OriginalPost) Url(baseURL string) string {
	return fmt.Sprintf("%s/posts/%d/%s", baseURL, i.Number, i.Slug)
}

File: public/models/post.ts
import { User } from "./identity"

export interface Post {
  id: number
  number: number
  slug: string
  title: string
  description: string
  createdAt: string
  lastActivityAt: string
  status: string
  user: User
  voteType: number
  response: PostResponse | null
  votesCount: number
  commentsCount: number
  tags: string[]
  tagDates?: string
  lockedSettings?: PostLockedSettings
  archivedSettings?: PostArchivedSettings
  upvotes?: number
  downvotes?: number
}

export function isPostLocked(post: Post): boolean {
  return !!post.lockedSettings && post.lockedSettings.locked;
}

export function isPostArchived(post: Post): boolean {
  return post.status === "archived";
}

export class PostStatus {
  constructor(public title: string, public value: string, public show: boolean, public closed: boolean, public filterable: boolean) {}

  public static Open = new PostStatus("Open", "open", false, false, true)
  public static Planned = new PostStatus("Planned", "planned", true, false, true)
  public static Started = new PostStatus("Started", "started", true, false, true)
  public static Completed = new PostStatus("Completed", "completed", true, true, true)
  public static Declined = new PostStatus("Declined", "declined", true, true, true)
  public static Duplicate = new PostStatus("Duplicate", "duplicate", true, true, true)
  public static Deleted = new PostStatus("Deleted", "deleted", false, true, true)
  public static Archived = new PostStatus("Archived", "archived", true, false, true)

  public static Get(value: string): PostStatus {
    for (const status of PostStatus.All) {
      if (status.value === value) {
        return status
      }
    }
    throw new Error(`PostStatus not found for value ${value}.`)
  }

  public static All = [PostStatus.Open, PostStatus.Planned, PostStatus.Started, PostStatus.Completed, PostStatus.Duplicate, PostStatus.Declined, PostStatus.Archived]
}

export interface PostLockedSettings {
  locked: boolean;
  lockedAt: string;
  lockedBy: User;
  lockMessage?: string;
}

export interface PostArchivedSettings {
  archivedAt: string;
  archivedBy: User;
  previousStatus: string;
}

export interface PostResponse {
  user: User
  text: string
  respondedAt: Date
  original?: {
    number: number
    title: string
    slug: string
    status: string
  }
}

export interface ReactionCount {
  emoji: string
  count: number
  includesMe: boolean
}

export interface Comment {
  id: number
  content: string
  createdAt: string
  user: User
  attachments?: string[]
  reactionCounts?: ReactionCount[]
  editedAt?: string
  editedBy?: User
}

export interface Tag {
  id: number
  slug: string
  name: string
  color: string
  isPublic: boolean
}

export interface Vote {
  createdAt: Date
  user: {
    id: number
    name: string
    email: string
    avatarURL: string
  }
}
